### DISCUSSION TITLE
Is there a EndPlayback callbeck for player?

### DISCUSSION THREAD

**Discussion by 0x0fe** [pschatzmann/arduino-audio-tools#757] [Q&A] [answered]
I cannto find any weakly defined callback for player "endPlayback". Since i reimplemented the files handling callback (due to indexing issue) until now i implemented a callback here, but it is only usable when we are in autoNext mode. So i wonder if there is any callback somewhere that will get called at the end of playback, no matter if we are in "autoNext" mode or not.

**Comment by pschatzmann** [comment, marked as answer]
There are many ways to solve this. The most obvious ones are:
- check the result from the copy() call. You will get 0 immediately if there is no more data.
- you can query the player if it is still active: e.g with if(player.isActive()) or just if(player). This is set to false after the timeout.
- you can create a subclass of AudioPlayer and overwrite the moveToNextFileOnTimeout() method to implement your own callback
- you can add your own callback method to the AudioPlayer class and share your work with a pull request
- leave the player in autonext mode and handle the logic in your nextStream() method of your custom audio source.

**Reply by 0x0fe** [reply]
regarding the EOF (end of file) callback, i really need to implement one now, which place would be the most appropriate?
in moveToNextFileOnTimeout after fade.setFadeInActive
or 
in writeEnd after p_decoder->begin();

**Comment by pschatzmann** [comment]
I don't really have any preference, so I let you decide. 
Is this supposed to be called after each file or after the last file ?

**Reply by 0x0fe** [reply]
after each file because this callback resets the sleep timer, wich shuts the device off after x minutes of inactivity. For now i tried to add it in moveToNextFileOnTimeout, testing.

**Comment by 0x0fe** [comment]
So, here is the EOF callback, so far it works fine.
By the way, when the player need to switch between different sources on the fly, do we have to keep all sources instanciated? i fear for the IRAM waste, being already very very tight.

`player.setAudioSource`


```
    virtual void setEofCallback(void (*callback)(void)) {
        eofcallback=callback;
    }

    protected:
        void (*eofcallback)(void);
```

```
        void moveToNextFileOnTimeout(){
            if (p_final_stream->availableForWrite()==0) return;
            if (p_input_stream == nullptr || millis() > timeout) {
                fade.setFadeInActive(true);
                eofcallback();
                if (autonext) {
                    LOGW("-> timeout - moving by %d", steam_increment);
                    // open next stream
                    if (!next(steam_increment)) {
                        LOGD("stream is null");
                    }
                } else {
                    active = false;
                }
                timeout = millis() + p_source->timeoutAutoNext();
            }
        }
```

**Comment by pschatzmann** [comment]
Will you submit a pull request ?
I would expect that only the actual source needs to be available in memory. Any other you could delete if you manage to restore the relevant state somehow if you need it

**Reply by 0x0fe** [reply]
> Will you submit a pull request ?

no, i work on the local source directly, that is why i copied it here.

> only the actual source needs to be available in memory

Ok noted, i will do that.

**Reply by pschatzmann** [reply]
Looking at the lines that you have deleted, I do not feel comfortable to use your suggested  code.
I propose that you add the deleted lines back or that you create your own subclass. I changed the moveToNextFileOnTimeout() to be virtual...
